src/hotspot/share/logging/logAsyncWriter.cpp

```diff
@@ -42,28 +43,65 @@ class AsyncLogWriter::AsyncLogLocker : public StackObj {
   }
 };
 
-void AsyncLogWriter::enqueue_locked(const AsyncLogMessage& msg) {
-  if (_buffer.size() >= _buffer_max_size) {
+// LogDecorator::None applies to 'constant initialization' because of its constexpr constructor.
+const LogDecorations& AsyncLogWriter::None = LogDecorations(LogLevel::Warning, LogTagSetMapping<LogTag::__NO_TAG>::tagset(),
+                                      LogDecorators::None);
+
+// Reserve space for a flush token, so 'push_flush_token' always succeeds.
+AsyncLogWriter::Buffer::Buffer(size_t capacity) : _pos(0) {
+  _buf = NEW_C_HEAP_ARRAY(char, capacity, mtLogging);
+  _capacity = capacity - AsyncLogWriter::TOKEN_SIZE;
+}
+
+AsyncLogWriter::Buffer::~Buffer() {
+  FREE_C_HEAP_ARRAY(char, _buf);
+}
+
+bool AsyncLogWriter::Buffer::push_back(LogFileStreamOutput* output, const LogDecorations& decorations, const char* msg) {
+  size_t len = strlen(msg) + 1; // including trailing zero
+  size_t sz = align_up(sizeof(Message) + len, sizeof(void*));
+
+  if (_pos + sz <= _capacity || output == nullptr/*token*/) {
+    new(_buf + _pos) Message(output, decorations, msg);
+    _pos += sz;
+    return true;
+  }
+
+  return false;
+}
+
+void AsyncLogWriter::Buffer::push_flush_token() {
+  bool result = push_back(nullptr, AsyncLogWriter::None, "");
+  assert(result, "fail to enqueue the flush token");
+}
+
+AsyncLogWriter::Message* AsyncLogWriter::Buffer::Iterator::next() {
+  assert(_curr < _buf._pos, "sanity check");
+  auto msg = reinterpret_cast<Message*>(_buf._buf + _curr);
```

Comment:
Is there a reason why Iterator needs to directly access _buf? Why not give Buffer a `const char* base() const` ? Also, why return a non-const ptr?